arch/RISCV/RISCVMapping.c

```diff
@@ -134,51 +514,132 @@ const char *RISCV_reg_name(csh handle, unsigned int reg)
 }
 
 static const insn_map insns[] = {
-	// dummy item
-	{
-	 0, 0,
-#ifndef CAPSTONE_DIET
-	 {0}, {0}, {0}, 0, 0
-#endif
-	 },
-
-#include "RISCVMappingInsn.inc"
+#include "RISCVGenCSMappingInsn.inc"
 };
 
 #ifndef CAPSTONE_DIET
 
 static const map_insn_ops insn_operands[] = {
-#include "RISCVMappingInsnOp.inc"
+#include "RISCVGenCSMappingInsnOp.inc"
 };
 
 #endif
 
-void RISCV_add_cs_detail(MCInst *MI, unsigned OpNum) {
+void RISCV_add_cs_detail_0(MCInst *MI, riscv_op_group opgroup, unsigned OpNum)
+{
+	printf("========== OP: %d, (CODE: %d)", OpNum, MI->Opcode);
+	switch (opgroup) {
+
+		case RISCV_OP_GROUP_Operand: printf("\n RISCV_OP_GROUP_Operand"); break;
+		case RISCV_OP_GROUP_BranchOperand:printf("\n RISCV_OP_GROUP_BranchOperand"); break;
+		case RISCV_OP_GROUP_VMaskReg:printf("\n RISCV_OP_GROUP_VMaskReg"); break;
+		case RISCV_OP_GROUP_VTypeI:printf("\n RISCV_OP_GROUP_VTypeI "); break;
+		case RISCV_OP_GROUP_ZeroOffsetMemOp:printf("\nRISCV_OP_GROUP_ZeroOffsetMemOp"); break;
+		case RISCV_OP_GROUP_Rlist:printf("\nRISCV_OP_GROUP_Rlist"); break;
+		case RISCV_OP_GROUP_Spimm:printf("\nRISCV_OP_GROUP_Spimm"); break;
+		case RISCV_OP_GROUP_CSRSystemRegister:printf("\n RISCV_OP_GROUP_CSRSystemRegister"); break;
+		case RISCV_OP_GROUP_RegReg:printf("\nRISCV_OP_GROUP_RegReg"); break;
+		case RISCV_OP_GROUP_FRMArg:printf("\nRISCV_OP_GROUP_FRMArg"); break;
+		case RISCV_OP_GROUP_FRMArgLegacy:printf("\nRISCV_OP_GROUP_FRMArgLegacy"); break;
+		case RISCV_OP_GROUP_FenceArg:printf("\nRISCV_OP_GROUP_FenceArg"); break;
+		case RISCV_OP_GROUP_FPImmOperand:printf("\nRISCV_OP_GROUP_FPImmOperand"); break;
+	}
+	printf("\n================================================================== %d\n", insn_operands[0].ops[0].type);
 	if (!detail_is_set(MI))
 		return;
-
-	cs_op_type op_type = map_get_op_type(MI, OpNum);
-
-	if (op_type == CS_OP_IMM) {
-		RISCV_get_detail_op(MI, 0)->type = RISCV_OP_IMM;
-		RISCV_get_detail_op(MI, 0)->imm = MCInst_getOpVal(MI, OpNum);
-		RISCV_get_detail_op(MI, 0)->access = map_get_op_access(MI, OpNum);
-		RISCV_inc_op_count(MI);
+	
+	cs_detail *details = MI->flat_insn->detail;
+	cs_riscv *riscv_details = &(details->riscv);
+	cs_riscv_op *op = &(riscv_details->operands[riscv_details->op_count]);
+	op->type = (riscv_op_type) map_get_op_type(MI, OpNum);
+	op->access = (cs_ac_type) map_get_op_access(MI, OpNum);
+
+	switch (map_get_op_type(MI, OpNum)) {
+		case CS_OP_REG:
+			op->reg = MCInst_getOperand(MI, OpNum)->RegVal;
+			printf("\n *******************  REG %d: %d\n", OpNum, op->reg);
+			break;
+		case CS_OP_MEM:
+			op->mem.base = 0;
+			op->mem.disp = MCInst_getOperand(MI, OpNum)->ImmVal;
+			printf("\n *******************  REG %d: %d\n", OpNum, op->reg);
+			break;
+		case CS_OP_IMM:
+			op->imm = MCInst_getOperand(MI, OpNum)->ImmVal;
+			printf("\n *******************  REG %d: %d\n", OpNum, op->reg);
+			break;
+		case CS_OP_MEM | CS_OP_REG:
+		case CS_OP_MEM | CS_OP_IMM:
+			break; // handle the weird combination later 
```

Comment:
This is not a weird combination. It indicates that the immediate is part of a memory operand. Like the case we discussed for `2(base_reg)`. The immediate `2` should have the type `CS_OP_MEM | CS_OP_IMM`, and `base_reg` should have the type `CS_OP_MEM | CS_OP_REG`. If they don't the TableGen definitions are weirdly defined or a case is missing.